chain/src/quarantined.rs

```diff
@@ -0,0 +1,371 @@
+    type IntoIter = <BTreeMap<IdentityKey, Unbonding> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.scheduled.into_iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a Scheduled {
+    type Item = <&'a BTreeMap<IdentityKey, Unbonding> as IntoIterator>::Item;
+    type IntoIter = <&'a BTreeMap<IdentityKey, Unbonding> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.scheduled.iter()
+    }
+}
+
+impl Extend<(IdentityKey, Unbonding)> for Scheduled {
+    fn extend<T: IntoIterator<Item = (IdentityKey, Unbonding)>>(&mut self, iter: T) {
+        for (identity_key, unbonding) in iter {
+            let entry = self.scheduled.entry(identity_key).or_default();
+            entry.extend(unbonding.note_payloads);
+            entry.extend(unbonding.nullifiers);
+        }
+    }
+}
+
+impl FromIterator<(IdentityKey, Unbonding)> for Scheduled {
+    fn from_iter<T: IntoIterator<Item = (IdentityKey, Unbonding)>>(iter: T) -> Self {
+        let mut scheduled = Scheduled::default();
+        scheduled.extend(iter);
+        scheduled
+    }
+}
+
+impl Protobuf<pb::quarantined::Scheduled> for Scheduled {}
+
+impl TryFrom<pb::quarantined::Scheduled> for Scheduled {
+    type Error = anyhow::Error;
+
+    fn try_from(value: pb::quarantined::Scheduled) -> Result<Self, Self::Error> {
+        Ok(Self {
+            scheduled: value
+                .per_validator
+                .into_iter()
+                .map(
+                    |pb::quarantined::ValidatorEntry {
+                         identity_key,
+                         unbonding,
+                     }| {
+                        if let Some(identity_key) = identity_key {
+                            Ok((
+                                identity_key.try_into()?,
+                                unbonding
+                                    .map(TryInto::try_into)
+                                    .transpose()?
+                                    .unwrap_or_default(),
+                            ))
+                        } else {
+                            Err(anyhow::anyhow!("missing identity key"))
+                        }
+                    },
+                )
+                .collect::<Result<BTreeMap<_, _>, _>>()?,
+        })
+    }
+}
+
+impl From<Scheduled> for pb::quarantined::Scheduled {
+    fn from(value: Scheduled) -> Self {
+        Self {
+            per_validator: value
+                .scheduled
+                .into_iter()
+                .map(
+                    |(identity_key, unbonding)| pb::quarantined::ValidatorEntry {
+                        identity_key: Some(identity_key.into()),
+                        unbonding: if unbonding.is_empty() {
+                            None
+                        } else {
+                            Some(unbonding.into())
+                        },
+                    },
+                )
+                .collect(),
+        }
```

Comment:
It'd be nice to have a comment here that highlights the canonical serialization issue and how the code addresses it: we know that the items will always be written in order, because they're stored in a BTreeMap, which iterates in order.

However, this means that the `Ord` impl on the relevant types is consensus-critical, and that could be good to note..